pkg/nitro_enclave_cpu_plugin/device_plugin.go

```diff
@@ -0,0 +1,338 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package nitro_enclave_cpu_plugin
+
+import (
+	"errors"
+	"fmt"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s-ne-device-plugin/pkg/config"
+	"net"
+	"os"
+	"os/signal"
+	"path"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const (
+	deviceName                     = "nitro_enclaves_cpus"
+	devicePluginServerReadyTimeout = 10
+)
+
+var cpuIdCounter = 0
+
+type IPluginDefinitions interface {
+	socketPath() string
+	devicePath() string
+	resourceName() string
+}
+
+type NECPluginDefinitions struct {
+	IPluginDefinitions
+}
+
+func (n *NECPluginDefinitions) socketPath() string {
+	return pluginapi.DevicePluginPath + deviceName + ".sock"
+}
+
+func (n *NECPluginDefinitions) devicePath() string {
+	return "/dev/" + deviceName
+}
+
+func (n *NECPluginDefinitions) resourceName() string {
+	return "aws.ec2.nitro/" + deviceName
+}
+
+type IBasicDevicePlugin interface {
+	Start() error
+	Stop()
+}
+
+// NitroEnclavesDevicePlugin implements the Kubernetes device plugin API
+type NitroEnclavesCPUDevicePlugin struct {
+	dev  []*pluginapi.Device
+	pdef IPluginDefinitions
+
+	stop   chan interface{}
+	health chan *pluginapi.Device
+
+	server *grpc.Server
+
+	pluginapi.DevicePluginServer
+	IBasicDevicePlugin
+}
+
+// reads the number of offline cpus from /sys/devices/system/cpu/offline
+func determineAdvisableCPUs() (int, error) {
+	data, err := os.ReadFile("/sys/devices/system/cpu/offline")
+	if err != nil {
+		return 0, fmt.Errorf("error reading offline CPUs: %v", err)
+	}
+
+	// Handle empty/unknown case
+	content := strings.TrimSpace(string(data))
+	if content == "" || content == "-1" {
+		return 0, nil
+	}
+
+	total := 0
+	ranges := strings.Split(content, ",")
+
+	for _, r := range ranges {
+		parts := strings.Split(r, "-")
+		switch len(parts) {
+		case 1:
+			// Single CPU
+			total++
+		case 2:
+			// CPU range
+			start, err1 := strconv.Atoi(parts[0])
+			end, err2 := strconv.Atoi(parts[1])
+			if err1 != nil || err2 != nil {
+				return 0, fmt.Errorf("invalid CPU range: %s", r)
+			}
+			total += end - start + 1
+		default:
+			return 0, fmt.Errorf("malformed CPU range: %s", r)
+		}
+	}
+
+	return total, nil
+}
+
+func generateEnclaveCPUID() string {
+	ctr := cpuIdCounter
+	cpuIdCounter++
+	return deviceName + "_" + strconv.Itoa(ctr)
+}
+
+//func (necdp *NitroEnclavesCPUDevicePlugin) releaseResources() {
+//	necdp.server = nil
+//	os.Remove(necdp.pdef.socketPath())
+//}
+
```

Comment:
This can probably go if it is commented out.

Reply:
done